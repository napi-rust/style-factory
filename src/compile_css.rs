use crate::options::{get_parser_options, get_printer_options};
use lightningcss::bundler::{Bundler, FileProvider, SourceProvider};
use std::error::Error;
use std::path::{Path, PathBuf};
use std::sync::{Arc, Mutex};

pub struct CompileResult {
  pub css: String,
  pub dependencies: Vec<String>,
}

struct TrackingProvider {
  file_provider: FileProvider,
  dependencies: Arc<Mutex<Vec<String>>>,
}

impl TrackingProvider {
  pub fn new() -> Self {
    TrackingProvider {
      file_provider: FileProvider::new(),
      dependencies: Arc::new(Mutex::new(Vec::new())),
    }
  }
}

impl SourceProvider for TrackingProvider {
  type Error = <FileProvider as SourceProvider>::Error;

  fn read(&self, path: &Path) -> Result<&str, Self::Error> {
    let result = self.file_provider.read(path)?;
    self
      .dependencies
      .lock()
      .unwrap()
      .push(path.to_string_lossy().to_string());
    Ok(result)
  }

  fn resolve(&self, specifier: &str, originating_file: &Path) -> Result<PathBuf, Self::Error> {
    self.file_provider.resolve(specifier, originating_file)
  }
}

pub fn compile_css(entry: &Path) -> Result<CompileResult, Box<dyn Error>> {
  let fs = TrackingProvider::new();
  let mut bundler = Bundler::new(&fs, None, get_parser_options());
  let stylesheet = bundler.bundle(entry).unwrap();
  let result = stylesheet.to_css(get_printer_options())?;
  let dependencies = fs.dependencies.lock().unwrap().clone();

  Ok(CompileResult {
    css: result.code,
    dependencies,
  })
}

#[cfg(test)]
mod tests {
  use super::*;
  use indoc::indoc;
  use insta::assert_snapshot;
  use std::fs;
  use tempfile::tempdir;

  #[test]
  fn test_basic_bundle() -> Result<(), Box<dyn Error>> {
    let dir = tempdir()?;
    let css_path = dir.path().join("a.css");

    fs::write(
      &css_path,
      indoc! { r#"
      @import "./b.css";
      .a { color: red; .a-child { color: blue; } }
    "#},
    )?;

    let png_data: &[u8] = &[
      0x89, 0x50, 0x4E, 0x47, 0x0D, 0x0A, 0x1A, 0x0A, 0x00, 0x00, 0x00, 0x0D, 0x49, 0x48, 0x44,
      0x52, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x01, 0x08, 0x02, 0x00, 0x00, 0x00, 0x90,
      0x77, 0x53, 0xDE, 0x00, 0x00, 0x00, 0x0C, 0x49, 0x44, 0x41, 0x54, 0x08, 0xD7, 0x63, 0xF8,
      0xFF, 0xFF, 0x3F, 0x00, 0x05, 0xFE, 0x02, 0xFE, 0xDC, 0xCC, 0x59, 0xE7, 0x00, 0x00, 0x00,
      0x00, 0x49, 0x45, 0x4E, 0x44, 0xAE, 0x42, 0x60, 0x82,
    ];
    fs::write(dir.path().join("b.png"), png_data)?;

    fs::write(
      dir.path().join("b.css"),
      indoc! { r#"
      @import "./c.css";
      .b { padding: 0; background: url("./b.png"); }
    "#},
    )?;

    fs::write(
      dir.path().join("c.css"),
      indoc! { r#"
      .c { padding: 0 }
    "#},
    )?;

    let result: CompileResult = compile_css(css_path.as_path())?;
    let dependencies = result.dependencies;

    assert_snapshot!(result.css);
    assert_eq!(dependencies.len(), 3);
    let dependencies_names: Vec<String> = dependencies
      .iter()
      .map(|path| path.split('/').last().unwrap().to_string())
      .collect();
    assert_eq!(dependencies_names, vec!["a.css", "b.css", "c.css"]);

    Ok(())
  }
}
