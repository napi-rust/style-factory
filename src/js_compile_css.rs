use crate::compile_css::compile_css;
use std::path::Path;

#[napi(object)]
pub struct JSCompileResult {
  pub css: String,
  pub dependencies: Vec<String>,
}

#[napi(js_name = "compileCSS")]
pub fn js_compile_css(entry: String) -> Result<JSCompileResult, napi::Error> {
  let entry = Path::new(&entry);
  let result = compile_css(entry);

  match result {
    Ok(result) => Ok(JSCompileResult {
      css: result.css,
      dependencies: result.dependencies,
    }),
    Err(err) => Err(napi::Error::new(
      napi::Status::GenericFailure,
      format!("Error: {}", err),
    )),
  }
}

#[cfg(test)]
mod tests {
  use super::*;
  use indoc::indoc;
  use insta::assert_snapshot;
  use std::fs;
  use tempfile::tempdir;

  #[test]
  fn test_js_compile_css() -> Result<(), Box<dyn std::error::Error>> {
    let dir = tempdir().unwrap();
    let entry = dir.path().join("entry.css");
    fs::write(
      &entry,
      indoc! {"
      @import 'foo.css';
      body {
        color: red;
      }
    "},
    )?;
    let foo = dir.path().join("foo.css");
    fs::write(&foo, "p { color: blue; }")?;

    let result = js_compile_css(entry.to_string_lossy().to_string()).unwrap();
    let dependencies = result.dependencies;
    assert_eq!(dependencies.len(), 2);
    assert_snapshot!(result.css);
    let dependencies_names = dependencies
      .iter()
      .map(|d| Path::new(d).file_name().unwrap().to_str().unwrap())
      .collect::<Vec<_>>();
    assert_eq!(dependencies_names, vec!["entry.css", "foo.css"]);
    Ok(())
  }
}
