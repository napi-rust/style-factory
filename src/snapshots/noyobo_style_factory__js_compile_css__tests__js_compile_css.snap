---
source: src/js_compile_css.rs
expression: result.css
---
p{color:#00f}body{color:red}
